examples_tests/61.UI/UI_System.cpp

```diff
@@ -0,0 +1,1102 @@
+
+		adjustGlobalFontScale();
+
+		updateDescriptorSets();
+
+		auto & io = ImGui::GetIO();
+		io.DisplaySize = ImVec2(window->getWidth(), window->getHeight());
+		io.DisplayFramebufferScale = ImVec2(1.0f, 1.0f);
+
+		state->vertexBuffers.resize(maxFramesInFlight);
+		state->indexBuffers.resize(maxFramesInFlight);}
+
+	//-------------------------------------------------------------------------------------------------
+
+	static void UpdateMousePositionAndButtons()
+	{
+		// TODO: Handle keyboard events
+		//// Set OS mouse position if requested (rarely used, only when ImGuiConfigFlags_NavEnableSetMousePos is enabled by user)
+		//if (io.WantSetMousePos)
+		//{
+		//    RF::WarpMouseInWindow(static_cast<int32_t>(io.MousePos.x), static_cast<int32_t>(io.MousePos.y));
+		//}
+		//else
+		//{
+		//    io.MousePos = ImVec2(-FLT_MAX, -FLT_MAX);
+		//}
+		//int mx, my;
+		//uint32_t const mouse_buttons = MSDL::SDL_GetMouseState(&mx, &my);
+		//io.MouseDown[0] = (mouse_buttons & SDL_BUTTON(SDL_BUTTON_LEFT)) != 0;  // If a mouse press event came, always pass it as "mouse held this frame", so we don't miss click-release events that are shorter than 1 frame.
+		//io.MouseDown[1] = (mouse_buttons & SDL_BUTTON(SDL_BUTTON_RIGHT)) != 0;
+		//io.MouseDown[2] = (mouse_buttons & SDL_BUTTON(SDL_BUTTON_MIDDLE)) != 0;
+		//if (RF::GetWindowFlags() & MSDL::SDL_WINDOW_INPUT_FOCUS)
+		//{
+		//    io.MousePos = ImVec2(static_cast<float>(mx), static_cast<float>(my));
+		//}
+
+		CommonAPI::InputSystem::ChannelReader<IMouseEventChannel> mouse;
+		//CommonAPI::InputSystem::ChannelReader<IKeyboardEventChannel> keyboard;
+
+		state->inputSystem->getDefaultMouse(&mouse);
+		//inputSystem->getDefaultKeyboard(&keyboard);
+
+		static std::chrono::microseconds previousEventTimestamp {};
+
+		mouse.consumeEvents([](const IMouseEventChannel::range_t& events) -> void {
+			auto& io = ImGui::GetIO();
+			for (auto event : events)
+			{
+				if (event.timeStamp < previousEventTimestamp)
+				{
+					continue;
+				}
+				previousEventTimestamp = event.timeStamp;
+
+				if(event.type == nbl::ui::SMouseEvent::EET_CLICK)
+				{
+					int buttonIndex = -1;
+					if (event.clickEvent.mouseButton == EMB_LEFT_BUTTON) 
+					{
+						buttonIndex = 0;
+					} else if (event.clickEvent.mouseButton == EMB_RIGHT_BUTTON)
+					{
+						buttonIndex = 1;
+					} else if (event.clickEvent.mouseButton == EMB_MIDDLE_BUTTON)
+					{
+						buttonIndex = 2;
+					}
+
+					if(event.clickEvent.action == SMouseEvent::SClickEvent::EA_PRESSED) {
+						io.MouseDown[buttonIndex] = true;
+					} else if (event.clickEvent.action == SMouseEvent::SClickEvent::EA_RELEASED) {
+						io.MouseDown[buttonIndex] = false;
+					}
+				}
+			}
+		});
+
+		auto const position = state->window->getCursorControl()->getPosition();
+		auto& io = ImGui::GetIO();
+		io.MousePos.x = static_cast<float>(position.x) - state->window->getX();
+		io.MousePos.y = static_cast<float>(position.y) - state->window->getY();
```

Comment:
I'd rather get the mouse position from an event, reason is that events distinguish between mice... cursor control doesn't (we should probably remove that method).